Mathlib/Data/Finset/Image.lean

```diff
@@ -657,6 +660,56 @@ theorem biUnion_singleton {f : α → β} : (s.biUnion fun a => {f a}) = s.image
 
 end Image
 
+/-! ### filterMap -/
+
+section FilterMap
+
+/-- `filterMap f s` is a combination filter/map operation on `s`.
+  The function `f : α → Option β` is applied to each element of `s`;
+  if `f a` is `some b` then `b` is included in the result, otherwise
+  `a` is excluded from the resulting finset.
+
+  In notation, `filterMap f s` is the finset `{b : β | ∃ a, f a = some b}`. -/
+-- TODO: should there be `filterImage` too?
+def filterMap (f : α → Option β) (s : Finset α)
+    (f_inj : ∀ a a' b, b ∈ f a → b ∈ f a' → a = a') : Finset β :=
+  ⟨s.val.filterMap f, s.nodup.filterMap f f_inj⟩
+
+@[simp]
+theorem filterMap_val (f : α → Option β) (s : Finset α)
+    {f_inj : ∀ a a' b, b ∈ f a → b ∈ f a' → a = a'} :
+    (filterMap f s f_inj).1 = s.1.filterMap f :=
+  rfl
+
+@[simp]
+theorem filterMap_empty (f : α → Option β) {f_inj : ∀ a a' b, b ∈ f a → b ∈ f a' → a = a'} :
+    (∅ : Finset α).filterMap f f_inj = ∅ :=
+  rfl
+
+variable (f : α → Option β) {s t : Finset α} {f_inj : ∀ a a' b, b ∈ f a → b ∈ f a' → a = a'}
+
+@[simp]
+theorem mem_filterMap {b : β} : b ∈ s.filterMap f f_inj ↔ ∃ a ∈ s, f a = some b :=
+  s.val.mem_filterMap f
+
+-- TODO: This was added in analogy to `coe_map`, but is this `simp` lemma even good?
+-- It might be nicer if it coerced `PFun.image f s` instead, but that's ill-typed because
+-- `f` is not a `PFun`.
+@[simp, norm_cast]
+theorem coe_filterMap : (s.filterMap f f_inj : Set β) = {b | ∃ a ∈ s, f a = some b} :=
+  Set.ext (by simp only [mem_coe, mem_filterMap, Option.mem_def, Set.mem_setOf_eq, implies_true])
+
+@[simp]
+theorem filterMap_some : s.filterMap some Option.mem_some_eucl = s :=
```

Comment:
Just make the proof an implicit argument of the theorem here: it doesn't matter which proof appears here!

Reply:
I don't agree, making it an argument means this can't be used to rewrite backwards

Reply:
```suggestion
theorem filterMap_some : s.filterMap some (by simp)  = s :=
```

Reply:
Changed to `(by simp)`